crates/vhost-user-backend/src/lib.rs

```diff
@@ -178,6 +190,39 @@ where
         }
     }
 
+    /// Bind to socket, handle a single connection and shutdown
+    ///
+    /// This is a convenience function that provides an easy way to handle the
+    /// following actions without needing to call the low-level functions:
+    /// - Create a listener
+    /// - Start listening
+    /// - Handle a single event
+    /// - Send the exit event to all handler threads
+    ///
+    /// *Note:* See [VhostUserDaemon::start] and [VhostUserDaemon::wait] if you
+    /// need more flexibility.
+    pub fn serve<P: AsRef<Path>>(&mut self, socket: P) -> Result<()> {
+        let listener = Listener::new(socket, true).map_err(Error::CreateVhostUserListener)?;
+
+        self.start(listener)?;
+        let result = self.wait();
+
+        // Regardless of the result, we want to signal worker threads to exit
+        self.handler.lock().unwrap().send_exit_event();
+
+        // For this convenience function we are not treating certain "expected"
+        // outcomes as error. Disconnects and partial messages can be usual
+        // behaviour seen from quitting guests.
+        match &result {
+            Err(e) => match e {
+                Error::HandleRequest(VhostUserError::Disconnected) => Ok(()),
+                Error::HandleRequest(VhostUserError::PartialMessage) => Ok(()),
```

Comment:
> It might be appropriate to let the caller decide what to do with the result.

Hm... I intentionally added it here since every daemon in vhost-device checked for this and logged a not very helpful message... But I got no strong opinions here. If this only happens rarely, maybe we can also just return an error and tweak the error string a little bit? @vireshk do you see this message often?

Reply:
IMO it doesn't matter if you see it often or not, it's a matter of letting the backend implementation decide if this is relevant or not.

Reply:
I understand. But there's a difference between factorizing code and hiding something from the caller.
I'll let the other reviewers chime in on that topic, I'm okay with the current code if everybody else is okay with it.

Reply:
> hiding something from the caller

Sure! We should only do this if we are sure that these are usually innocent errors. Maybe there is some way to check whether the other end actually hung up after receiving a partial message...

Reply:
Any other opinions from other CODEOWNERS? :)

Reply:
That might be fine with me, but I would specify it better in the documentation that these errors are ignored as they are expected during frontend closure.

An alternative I can think of is to add a parameter to ignore them or not, but I'm not sure, in the end, if a backend wants to have complete control, it can use the functions that are already there.

Reply:
I will mention it in the docs.
I would like to avoid an additional parameter. My goal was to cover the 90% case with 10% of the code. If that does not work this way, I would go back to the drawing board.

Reply:
@sboeuf: Does the update with explicitly documenting this works for you?

Reply:
Yes looks good

Reply:
Are you happy to review the PR then? :innocent: